---
title: "Lab Talk - Riboviz - Putting the 'viz' back into 'riboviz': connecting Rmarkdown and Nextflow"
author: "Flic Anderson"
date: "10/05/2021"
output: 
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/Rmarkdown-and-output"></footer>');    
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 100px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>


# *Putting the 'viz' back into 'riboviz':* connecting Rmarkdown and Nextflow

<div class="notes">

Bio: 

Flic Anderson is a Research Assistant in Bioinformatics in the Wallace Lab (within the Institute for Cell Biology) at the University of Edinburgh. The Wallace Lab works to understand how translation works within cells (particularly of fungi) and what factors control and regulate protein synthesis. 

Flic is part of the development team for the 'riboviz' open source software package for processing and analysis of ribosome profiling data, working with colleagues from EPCC (University of Edinburgh) to make the codebase more robust & sustainable.   

With a background in Ecology and Botany (previously a Research Assistant at the Royal Botanic Garden Edinburgh), Flic has worked with a wide range of types of biological data using different software tools and programming languages (particularly R) for the last 8 years, and found herself drawn towards bioinformatics and research software engineering. 

Flic is a member of the Edinburgh Carpentries and a certified Carpentries instructor in foundational coding and data science skills for researchers. 


Title: 

Putting the 'viz' back into 'riboviz': connecting Rmarkdown, Shiny and Nextflow


Abstract:  

'riboviz' is an open source software package for processing and analysing ribosome profiling (RNA sequence) data.

This talk will briefly cover the implementation of an Rmarkdown HTML output report and Shiny app for data visualisation of ribosome profiling datasets, added as part of development efforts to make riboviz more user-friendly and effective.  

Flic will describe the initial approach and pitfalls encountered while developing the new visualisation features, and discuss integrating the visualisation side of riboviz with the recently implemented workflow management system (Nextflow) riboviz has moved to.  

Including a short primer on what Workflow Management Systems such as Nextflow 'do' and why they're useful for complex data pipelines, and details on how to get Rmarkdown HTML reports to play nicely as part of a Nextflow workflow. 


</div>


## HOW TO (RE-)USE THIS MATERIAL

This is a `.html` presentation created in `R Markdown` with `ioslides`. 

(It's been written in a [.Rmd](http://rmarkdown.rstudio.com) file, and we generated .html slides by 'knitting' it in Rstudio.)

You can check out the code used to make these slides at the [Workshop repo](https://github.com/FlicAnderson/20201127-gitflow-workshop) on Github, and adapt it for your own presentations if you like - we've got a MIT Licence on the repo, which means: 

*"Basically, you can do whatever you want as long as you include the original copyright and license notice in any copy of the software/source."*

Source: [tl;drLegal](https://tldrlegal.com/license/mit-license)

# Outline: 

## Outline: 

* riboviz & ribosome profiling refresher
* what on earth is a WfMS? 
* Nextflow in a nutshell  
* Rmarkdown and out...(puts)
* what I tried (aka the looping horror)
* what actually worked... 
* what next? 

# Intro:

## RiboViz

**Riboviz processes & analyses ribosome profiling data**  

Ribosome profiling data helps unlock details of **active translation**:

  * *mechanics of translation*  
  * *regulation methods*   
  * *translational efficiency*  

Developing/improving riboviz = **more researcher time** for biological questions

Robust & reliable code

<div class="notes">

more researcher time: rather than tinkering with pipelines & bespoke analysis code…  

Robust & reliable code gives more confidence in results

</div>

## Features of Ribosome Profiling Data

* Looking for **3-nucleotide periodicity**: ribosomes moving along mRNA transcript 1 codon (3 nucleotides) at a time  

* Most reads **map to coding regions** (98.8% in Ingolia et al. 2009)  
* Reasonable read-lengths (e.g. should look for **~28-30NT**)  

* Looking for most reads to be **in one frame**  


<font size="2">Ingolia et al. 2009 - "Genome-wide analysis in vivo of translation with nucleotide resolution using ribosome profiling", doi: 10.1126/science.1168978</font>

<div class="notes">

Ingolia et al. 2009 - Science 324(5924):218-23 "Genome-wide analysis in vivo of translation with nucleotide resolution using ribosome profiling", doi: 10.1126/science.1168978.

28-30NT due to the 'width' of the ribosome as this is what 'protected' the mRNA from the nucleases in the experiment's library preparation steps.

</div>

## RPF Data & Reading Frame 


<div class="centered">
```{r, out.width = "700"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-output_frame.png")
``` 
</div>

Many riboviz output plots let us examine these features.  


<font size="2">Reading Frame: https://www.ncbi.nlm.nih.gov/Class/MLACourse/Original8Hour/Genetics/readingframe.html; Data: Guydosh & Green (2014). doi: 10.1016/j.cell.2014.02.006.</font>


<div class="notes">

Ribosome profiling data is really useful for investigating translational mechanics particularly because of the information it provides about the active open reading frame!

Because a codon is 3 nucleotides, where a ribosome 'starts' reading the mRNA will affect what's translated. 

Can locate the start codon ‘AUG’ (from annotation position information from a .gff or sequence info) & determine what should be the reading frame based on that. 

Can determine which reading frame(s) are being actively translated from the counts of reads sequenced which match each frame. Can look at if this is altered by particular characteristics of the gene or sequence for example.

Outputs like these help users determine whether the data looks as expected (assists QC).  

Data: Guydosh & Green (2014). “Dom34 rescues ribosomes in 3' untranslated regions.” Cell. 156 (5): 950-62. doi: 10.1016/j.cell.2014.02.006.   

</div>




## Ribosome Profiling Method

<div class="centered">
```{r, out.width = "650"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/Ingolia2014_polysomes-to-footprints.png")
``` 
</div>
<font size="2">Ingolia et al. 2009</font>

<div class="notes">

Ribosome Profiling Method:
Lyse the cells to get at the mRNA molecules bound to ribosomes.  
'Stop' the translation process: e.g. with cycloheximide or other means   
Digest the non-protected RNA using a nuclease  
Strip away the ribosomes and proteins  
Size-select for these previously 'masked' fragments of mRNA  
Add adapters  
Reverse-transcribe to complimentary DNA  
Amplify  
Sequence   

Take these reads & submit to further processing & analysis to get plots & information.

That’s where riboviz comes in!

Processing:
Multiple steps using a range of other tools ‘plumbed-in’.

Analysis:
Analyse & quantify data: 
Create outputs (including for quality-control, further analysis)

Ingolia NT, Ghaemmaghami S, Newman JR, Weissman JS (2009). "Genome-wide analysis in vivo of translation with nucleotide resolution using ribosome profiling". Science. 324 (5924): 218–23. doi:10.1126/science.1168978

</div>


## Riboviz Workflow: Inputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_inputs.png")
``` 
</div>

<div class="notes"> 

Additional organism-specific data files: ie the pipeline can run without them, but will not produce specific plots without them. 

</div>


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-aligned.png")
``` 
</div>

<div class="notes"> 

Processing:  
* Removing adapter sequences  
* Remove UMIs (Unique Molecular Identifiers) & barcodes if present  
* Demultiplex / Deduplicate reads if required  
* Need to filter out contaminant reads  
* Align reads to transcriptome   

</div>


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-ribogrid.png")
``` 
</div>

<div class="notes">

Storing data in really efficient .h5 format (hierarchical, like a file system w/ folders type of organisation)

</div>


## Riboviz Workflow: Analysis

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_analysis.png")
``` 
</div>

<div class="notes">

Analyse & quantify data:   
Create outputs (including for quality-control, further analysis)

One main R script which generates these plots & various analysis data; this is run for each sample, outputting .tsv and .pdf files into a folder for each sample within the overall dataset output folder.

</div>


## Riboviz Workflow: .PDF Outputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_outputs.png")
``` 
</div>  


<div class="notes">

Original outputs: .tsv data files and .pdf plot images for each sample. 

Dataset shown: B-Sc_2012: Saccharomyces cerevisiae yeast RPF samples at a range of meiotic phases from: High-resolution View of the Yeast Meiotic Program Revealed by Ribosome Profiling, Brar et al. 2012. https://doi.org/10.1126/science.1215110

Files present: ‘all’ - not all organisms have the same amount of data availale, and so information on tRNAs for example isn't available for all species, or isn't currently in a riboviz friendly format.  

This means we have some 'conditional' outputs. Not all taxa will have all of these outputs, which is something we've had to handle in the code. 

Yaml config file available at https://github.com/riboviz/example-datasets/blob/master/fungi/saccharomyces/Brar_2012_Meiosis_RPF_6-samples_CDS_w_250utrs_config.yaml

This was a dataset selected by Ania Kurowska (undergrad project in the Wallace lab, 2020)

</div>


## Riboviz Workflow: Visualization (NEW)

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>

## Workflow Management Systems

Workflow Management Systems (WfMS) provide infrastructure to set up, run & monitor a defined set of tasks (processes). 

Key features include: 

 * software management (can easily 'containerise' code & dependencies for easier deployment in different computing environments)  
 
 * portability / interoperability (the same workflow can be run in different computing infrastructure such as HPC or local machine) 
 
 * run time management (the WfMS handles the different tasks and manages their execution - instead of running 'in order', tasks can be split up and parallelised to run at the same time; error handling & logging generated by system)

* reproducibility (having the 'admin' handled by the WfMS means the same results on re-run, regardless of platform; specification is clearer)  

* communicable (can share your workflow more easily as it includes specification details as well as describing the workflow) 

* faster troubleshooting (many WfMS let you re-run certain steps of a workflow if something has changed or broken, without needing to re-run ALL parts)
 
Some examples are: Snakemake, Toil, cwltool and Nextflow


<div class="notes">

My colleagues involved on riboviz wrote a paper on how we selected the right workflow management system for the project: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008622  

These tasks might depend on each other, for example as in a data processing pipeline (where the output of one stage becomes the input of the next step) or allow parts of the pipeline to be re-run without needing to start from the beginning. 

A Workflow Management System lets you join up all the tasks into one workflow, which you can run with one command. No chopping & changing between software, languages, or different environments!


</div>



## Why not just use a script? riboviz as a case study

'riboviz' was previously run by a python script. Running this script then carried out all the data processing and analysis in bash, python and R. 

But as we added more functionality, extra tools and features, and it became more complex, we spent more time writing features around running the steps of the pipeline and less actual biological feature content!  

WfMS handle more of the process-related admin such as ensuring task inputs exist, creating log files, and queuing up the tasks to run in the right order and in an efficient manner according to the computing infrastructure. 

This latter point is key: in a script, tasks will happen sequentially based on their position. Having a WfMS splitting up these tasks and running them as soon as they're 'ready' makes it much more efficient (important if you're running big datasets through complex pipelines which can take hours!)

<div class="notes">

More details on riboviz and requirements for a WfMS in the paper: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008622  

Think of a script which runs lots of separate functions to do different data analysis tasks: you can execute the whole script, and all the commands inside are run. But, often workflow management systems allow several of those commands to be run at the same time, rather than waiting until their 'turn' in the script. 

</div>


## Nextflow

* parallelism - processes are run as soon as their required input is there - no hanging around for unrelated tasks to finish! 
* can run a pipeline on local computer vs compute cluster without tweaking it  
* lets you re-run the pipeline from the last successful point if it stops
* simple syntax so you can re-use existing scripts
* you can use containerisation (such as docker/singularity) and conda to manage your environment


<div class="notes">  

https://www.nextflow.io/

Find out more: Nextflow Carpentries style workshop (alpha!) lesson materials under development https://carpentries-incubator.github.io/workflows-nextflow/

Nextflow connects 'processes' (the steps or tasks) according to their outputs and inputs to other processes. 
 
Nextflow combines a **runtime environment** (the stuff that runs other stuff) and a **domain specific language** (a language which is specific to a particular type of task or activity, and not really intended for use outside it's area).

DSL: html is a DSL for web pages for example, SQL for databases. 


Answer: 
* what is it
* why do we use it?
* what was the reasoning of switching to this from python/bash?

</div>



## Rmarkdown 

R Markdown creates a notebook-style document 

Combine **text**, run R **code** (generate plots and calculate results) & easily integrate these into the document with **markdown formatting**.  

* Share code & conclusions with collaborators, who can then re-run and edit the analysis.  
* Can run code chunks in different languages (e.g. R, Python & SQL)  

* Can feed in parameters to .Rmd 

* Lots of output formats: .pdf, .html, word files, slideshows (like this one!)


<div class="notes">  

More info: https://r4ds.had.co.nz/r-markdown.html

R Markdown creates a notebook-style document which lets you combine text, run R code, and generate plots & results which are integrated into the document.  

Great for collaborating

Automatically-generated reports based on updated data (handy to show decision-makers), re-run with ease

Lots of output formats: .pdf, word files, slideshows (like this one!)

Benefits:  

* integrated w/ RStudio, so easy to develop & run  
* code 'chunks' let you show code, hide it, run it or not run it - very flexible  
* integrating code into text is very easy: (back-tick) r calculation-code (back-tick) 
* different language chunk ability is really interesting, versatile.
* YAML header/front-matter useful to feed in parameters to the .Rmd (e.g. as you might with Rscript on commandline); can set defaults in YAML

</div>


## Rmarkdown Example
<div class="centered">
```{r, out.width = "425"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/rmarkdown_code.png")
``` 
</div>

<div class="notes">

Features of Rmarkdown: 

* YAML header (surrounded by 3 dashes in a row above & below)   
* code blocks w/ run options (surrounded with back-ticks)  
* text blocks (not surrounded with anything)  
* markdown text formatting (use # for headings and _italics_ )

Can load required packages, do calculations in-text, generate & display plots.  

</div>

## Rmarkdown Example
<div class="centered">
```{r, out.width = "500"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/rmarkdown_html-output.png")
``` 
</div>


<div class="notes">  

Example based on https://r4ds.had.co.nz/r-markdown.html

</div>



## Riboviz HTML output 


Aiming for: all the plots for each sample in separate .html pages, to allow scrolling through & easier comparison.  

Keeping the plots for 1 sample on 1 page also makes it easier to keep track of which sample's outputs you're looking at & speeds up the QC process after a riboviz run has completed.

Also helps for communicating the outputs between colleagues - can just refer someone to 1 file, rather than multiple files or putting together a zip file.


<div class="notes">

* .pdf files for separate plots are useful, but having the same kind of plot (for example 3-nt periodicity) spread through separate folders means it's harder to compare between samples.  Aiming to improve it a little  - it would be easier to click between tabs at least!

* imagine having 4 or 5 pdf windows open, none of which have the sample_id (sample name) in their filename, or saved in the .pdf...

</div>


## Initial Attempts (DUPLICATION)

* Copied across the relevant calculation / plotting code into standard .Rmd code blocks
* meant twice as much to maintain, increased likelihood of errors or divergeance of code 
* inefficient to run calculation code twice - once in main analysis R script, then again in .Rmd call. 


<div class="notes">   

Wilson G, Aruliah DA, Brown CT, Chue Hong NP, Davis M, Guy RT, et al. (2014) Best Practices for Scientific Computing. PLoS Biol 12(1): e1001745. https://doi.org/10.1371/journal.pbio.1001745  

Don't repeat yourself (or others).:  
* Modularize code rather than copying and pasting.
* Re-use code instead of rewriting it.

</div>  

## Initial Attempts (CODE CHUNKS)

Running chunks of code from main analysis R script from within the .Rmd file via `knitr::read_chunk()`  

Passing parameters in via the YAML section again 
Could run via commandline - required step to be able to 'plumb' it into the Nextflow script   

* still inefficient, still running calculation code twice, but at least this time only from 1 location  
* slightly difficult to problem-solve if there were issues, because the .Rmd errors referred only to a line which referred to a block of code in .R file   
* difficult to ensure parameters were correct  


<div class="notes">  

Don't repeat yourself (or others).:  
* Every piece of data must have a single authoritative representation in the system.
* Re-use code instead of rewriting it.  


Looping through code blocks - but this ran calculation code twice - not efficient

Figuring out how to get the parameters passed in correctly from 'outside' + how to set default params.

How to run from commandline? This was needed before it'd be integratable w/ Nextflow  

</div>

## Initial Attempts (CODE CHUNKS)

<div class="centered">
```{r, out.width = "850"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/initial_code-chunks-rscript.png")
``` 
</div>

Main calculation & plot chunk, creates plot object three_nucleotide_periodicity_plot (rscripts/generate_stats_figs.R)


## Initial Attempts (CODE CHUNKS)

<div class="start">
```{r, out.width = "600"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/initial_code-chunks-rmarkdown_read-chunk.png")
``` 
</div> 

<div class="start">
```{r, out.width = "500"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/initial_code-chunks-rmarkdown_3NT.png")
``` 
</div>
<font size="2">
Line 149: tells script where we're reading chunks of code from  
Line 174: **running the code chunk** for 3nt periodicity to calculate & generate three_nucleotide_periodicity_plot  
Line 176: display plot 
</font>



<div class="notes">  

It's not immediately clear in the code what's happening, and no idea what's included in that block, so would need to trace it back in the other script. 

Could lead to errors if we can't see what's going on. 

Eventually realised that this was an unnecessarily difficult & complicated method of trying to achieve some plots in .html. 

</div>

## Next Attempts 

Create .tsv files containing all data objects required for plotting (previously we didn't have this for ALL plots) as part of main analysis R script    

Then loaded these .tsv files in the .Rmd & plot

Benefits:  
* not calculating anything in the .Rmd, as just need to load the .tsv files & plot  
* don't need to parse lots of parameters repeated from the config.yaml file, just the plot data .tsv files, main config yaml file name/location & current sample_id 


<div class="notes">
Decided to follow the Wilson et al 2014 principles: 

Don't repeat yourself (or others).:  
* Every piece of data must have a single authoritative representation in the system.
* Modularize code rather than copying and pasting.
* Re-use code instead of rewriting it. 

The main config yaml file is our 'single authoritative representation'.  
Let's modularise the code so we can just use the plot functions rather than needing to calculate, create an intermediate data object, and then plotting that.  
Re-using our original plotting functions!

</div>


## The Solution 

Edit main analysis script Nextflow process generateStatsFigs:  
 - create new .tsv files holding the data objects required to generate the output plots  

Create new Nextflow process for staticHTML task:   
 - staticHTML runs once per sample, passing config yaml file, sample_id information & the relevant plot data .tsv files into the .Rmd render commmand  
 - AnalysisOutputs.Rmd then checks which files names/locations have been passed in as parameters  
 - based on which files are present & on certain yaml parameters, the right .tsv files are then loaded and plots created on the .html page for that sample. 

Nextflow handles passing in the correct conditional files to the .Rmd render command, and that .Rmd script plots the correct 'conditional' plots by reading certain parameters in the main configuration script file.

 

<div class="notes">

Handling conditionals in Nextflow has been a little trickier than expected.    

staticHTML generates the text for the render command based on whether .tsv files are present or not, and contains the correct file paths (including the sample name folder)

</div>


## AnalysisOutputs.Rmd 

<div class="centered">
```{r, out.width = "1000"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/rmarkdown_final-yaml.png")
``` 
</div>  

<div class="centered">
```{r, out.width = "1000"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/rmarkdown_final-read-yaml.png")
``` 
</div>  

<div class="notes">

Line 5: specify document type (HTML)  
Line 6-9: table of contents  
Line 10: tells .Rmd which parameters to expect, and gives default values for each parameter  

Line 38: using {yaml} package, read contents of the config yaml file into yaml object; using 'params$' notation to pull this parameter info from what was fed into the .Rmd during the rmarkdown::render() command

</div>


## AnalysisOutputs.Rmd 

<div class="centered">
```{r, out.width = "1000"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/rmarkdown_final-load-tsv.png")
``` 
</div>  

<div class="centered">
```{r, out.width = "1000"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/rmarkdown_final-plot.png")
``` 
</div>  

* easy to read, easy to tell which functions are used 
* not repeating blocks of code or double-calculating  
* generates required output!

<div class="notes">

Line 127:  reading in the .tsv file holding data required for 3nt periodicity plot, pulling file info from parameter with 'param$'  

Line 186:  rmarkdown block for 3n5 plot: echo=FALSE argument for the block means "DON'T show the code, but DO run it"  

Line 189: use existing Plot function (loaded earlier in the .Rmd via source()) to plot the data loaded in from the .tsv

</div>


## staticHTML Nextflow process
<div class="centered">
```{r, out.width = "700"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/nextflow-process_staticHTML.png")
``` 
</div>  


<div class="notes">

The process is run once per sample listed in the config yaml file 

Line 643: starts the Nextflow process definition  
Line 644: sets output location for files  
Line 645-648: where the input files for the process get listed (these determine when the process is run)  
Line 650: tells Nextflow what the outputs of this process are (1x file w/ filename in this format)  
Line 652: this process is only run if run_static_html is TRUE in the main config yaml file   
Line 654-666: builds 'script' from: the rmarkdown::render() command which runs the Rmarkdown file & adds in the correct parameters (config file, sample_id, and the correct .tsv files according to conditionals set by the config file parameters)  
Line 668: actually executes the script (runs the render command w/ correct parameters)


process staticHTML {
    publishDir "${params.dir_out}/${sample_id}" 
    input:
      * config yaml file  
      * sample_id  
      * .tsv files for generating plots  
    output:
      file "${sample_id}_output_report.html"  
    when:
      params.run_static_html  
    shell:
      script = "rmarkdown::render('${projectDir}/rmarkdown/AnalysisOutputs.Rmd',"
      script += "params = list("
      script += "yamlfile='\$PWD/${config_file_yaml}', "
      script += "sampleid='!{sample_id}', "
      script += "three_nucleotide_periodicity_data_file = '$PWD/${sample_nt3_periodicity_tsv}', "
        (+ some other files)
      if (is_features_file) {
          script += ", sequence_features_file='$PWD/${sample_sequence_features_tsv}' "
      }
        (+ some other conditional files)
      script += "), "
      script += "output_format = 'html_document', "
      script += "output_file = '$PWD/${sample_id}_output_report.html')"
      """
      Rscript -e "${script}"
      """
} 

</div>



## HTML Report Example
<div class="centered">
```{r, out.width = "675"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-output_html-report-vignette.png")
``` 
</div> 

<div class="notes">

Shows the plot for 3-nucleotide periodicity  

Same as .pdf version previously output, but now have it clearly associated w/ provenance info, sample_id and date

</div>

## Summary / Lessons Learned

Although you CAN call chunks of code external to an R Markdown script, you probably shouldn't... 

Getting file paths right when passing files in as parameters to things is often tricky, but Nextflow makes it reasonably easy once you've got the 'knack'  

Working on adding this feature has been another reminder of how important it is to make sure your function only does 1 thing & is modular, as well as why it's vital to stick to "don't repeat yourself (or others!)"


<div class="notes">
... slide notes ...
</div>


## Thanks/Acknowledgements

riboviz is a BBSRC-NSF funded collaborative project: 


* **The Wallace Lab** - *University of Edinburgh*:  
Edward Wallace, Emma MacKenzie, Sophie Winterbourne + past students (Ania Kurowska, Siyin Xue)  
* **The Shah Lab** -  *Rutgers University*:  
Premal Shah, John Favate, Alex Cope, Peter Tilton  
* **The Lareau Lab** - *University of California, Berkeley*:  
Liana Lareau, Amanda Mok, Vikram Shivakumar  
* **EPCC** - *University of Edinburgh*:  
Kostas Kavousannakis, Mike Jackson, Junchen Liu  


<div class="notes">
... slide notes ...
</div>


# Questions?


<div class="notes">
... slide notes ...
</div> 




## Takeaway for pipeline builders: 

WfMS are good, and very useful 

Can be tricky to get your head around - another thing you 'have' to learn before your new feature will work!

If you realise you're spending more time on managing 'task admin' in your development than on research-focused features, you need a WfMS!


<div class="notes">
... slide notes ...
</div>


## Takeaway for lab folks:  

WfMS are good, and very useful, but if you're not building really complex stacks of code, don't worry too much about them. 

Instead, think smaller: 

You can get many of the same benefits (increased reproducibility, ease of use, shareability) with a script if you're not already using one. 

Data analysis is a great example: Rmarkdown documents for data exploration and analysis!

Tips for getting your code to work on your colleague's system! 
 - version control!
 - remember which packages/software are required & may need to be installed 
 - use package::function() in R :) 
 - dependencies might still sink you, but there is hope (containerisation, rEnv, being specific about which version of everything YOU used etc)  
 - be clear about file management (where do you expect the data files to be saved? did you write that down?)




<div class="notes">
... slide notes ...
</div>


## Shiny




<div class="notes">
... slide notes ...
</div>


## Riboviz Shiny output 




<div class="notes">
... slide notes ...
</div>



## Initial Attempts (Shiny)




<div class="notes">
... slide notes ...
</div>   

## The Solution (Shiny)




<div class="notes">
... slide notes ...
</div>
