---
title: "Lab Talk - Riboviz"
author: "Flic Anderson"
date: "10/05/2021"
output: 
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/Rmarkdown-and-output"></footer>');    
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 100px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>


# TITLE: Workflow management systems & setting up Rmarkdown reports

## HOW TO (RE-)USE THIS MATERIAL

This is a `.html` presentation created in `R Markdown` with `ioslides`. 

(It's been written in a [.Rmd](http://rmarkdown.rstudio.com) file, and we generated .html slides by 'knitting' it in Rstudio.)

You can check out the code used to make these slides at the [Workshop repo](https://github.com/FlicAnderson/20201127-gitflow-workshop) on Github, and adapt it for your own presentations if you like - we've got a MIT Licence on the repo, which means: 

*"Basically, you can do whatever you want as long as you include the original copyright and license notice in any copy of the software/source."*

Source: [tl;drLegal](https://tldrlegal.com/license/mit-license)

# Outline: 

## Outline: 

* option
* option
* option

# Intro:

## RiboViz

**Riboviz processes & analyses ribosome profiling data**  

Ribosome profiling data helps unlock details of **active translation**:

  * *mechanics of translation*  
  * *regulation methods*   
  * *translational efficiency*  

Developing/improving riboviz = **more researcher time** for biological questions

Robust & reliable code

<div class="notes">

more researcher time: rather than tinkering with pipelines & bespoke analysis code…  

Robust & reliable code gives more confidence in results

</div>

## Features of Ribosome Profiling Data

* Looking for 3-NT periodicity: ribosomes moving along transcript 1 codon at a time  

* Most reads map to coding regions (98.8% in Ingolia et al. 2009)  
* Reasonable read-lengths (e.g. know should look for appx ~28-30NT)  

* Looking for most reads to be in one frame

<div class="notes">
... slide notes ...
</div>

## RPF Data & Reading Frame 


<div class="centered">
```{r, out.width = "700"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/Ingolia2014_polysomes-to-footprints.png")
``` 
</div>

<div class="notes">
Reading Frame: https://www.ncbi.nlm.nih.gov/Class/MLACourse/Original8Hour/Genetics/readingframe.html

Data: Guydosh & Green (2014). “Dom34 rescues ribosomes in 3' untranslated regions.” Cell. 156 (5): 950-62. doi: 10.1016/j.cell.2014.02.006.

</div>




## Ribosome Profiling Method

<div class="centered">
```{r, out.width = "700"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/Ingolia2014_polysomes-to-footprints.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Inputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_inputs.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-aligned.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-ribogrid.png")
``` 
</div>

<div class="notes">
...
</div>


## Riboviz Workflow: Analysis

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_analysis.png")
``` 
</div>

<div class="notes">

One main R script which generates these plots & various analysis data; this is run for each sample, outputting .tsv and .pdf files into a folder for each sample within the overall dataset output folder.

</div>


## Riboviz Workflow: .PDF Outputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_outputs.png")
``` 
</div>  


<div class="notes">

Original outputs: .tsv data files and .pdf plot images for each sample. 

Dataset shown: B-Sc_2012: Saccharomyces cerevisiae yeast RPF samples at a range of meiotic phases from: High-resolution View of the Yeast Meiotic Program Revealed by Ribosome Profiling, Brar et al. 2012.

Files present: ‘all’ - not all organisms have the same amount of data availale, and so information on tRNAs for example isn't available for all species, or isn't currently in a riboviz friendly format.  

This means we have some 'conditional' outputs. Not all taxa will have all of these outputs, which is something we've had to handle in the code. 

Yaml config file available at https://github.com/riboviz/example-datasets

</div>


## Riboviz Workflow: Visualization (NEW)

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>

## Workflow Management Systems



<div class="notes">

My colleagues involved on riboviz wrote a paper on how we selected the right workflow management system for the project: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008622  



</div>



## Nextflow

* what is it
* why do we use it?
* what was the reasoning of switching to this from python/bash?


<div class="notes">
... slide notes ...
</div>


## Rmarkdown 




<div class="notes">
... slide notes ...
</div>



## Rmarkdown Outputs

Aiming for: all the plots for each sample in separate .html pages, to allow scrolling through & easier comparison.  

Keeping the plots for 1 sample on 1 page also makes it easier to keep track of which sample's outputs you're looking at & speeds up the QC process after a riboviz run has completed.

Also helps for communicating the outputs between colleagues - can just refer someone to 1 file, rather than multiple files or putting together a zip file.


<div class="notes">

* .pdf files for separate plots are useful, but having the same kind of plot (for example 3-nt periodicity) spread through separate folders means it's harder to compare between samples.  Aiming to improve it a little  - it would be easier to click between tabs at least!

* imagine having 4 or 5 pdf windows open, none of which have the sample_id (sample name) in their filename, or saved in the .pdf...

</div>




## Riboviz HTML output 




<div class="notes">
... slide notes ...
</div>


## Initial Attempts (HTML)




<div class="notes">

Looping through code blocks - but this ran calculation code twice - not efficient

Figuring out how to get the parameters passed in correctly from 'outside' + how to set default params.

How to run from commandline? This was needed before it'd be integratable w/ Nextflow

</div>



## Next Attempts 




<div class="notes">
... slide notes ...
</div>


## The Solution (HTML)




<div class="notes">
... slide notes ...
</div>




## Summary / Lessons Learned




<div class="notes">
... slide notes ...
</div>


## Thanks/Acknowledgements

riboviz is a BBSRC-NSF funded collaborative project: 


* **The Wallace Lab** - *University of Edinburgh*:  
Edward Wallace, Emma MacKenzie, Sophie Winterbourne + past students (Ania Kurowska, Siyin Xue)  
* **The Shah Lab** -  *Rutgers University*:  
Premal Shah, John Favate, Alex Cope, Peter Tilton  
* **The Lareau Lab** - *University of California, Berkeley*:  
Liana Lareau, Amanda Mok, Vikram Shivakumar  
* **EPCC** - *University of Edinburgh*:  
Kostas Kavousannakis, Mike Jackson, Junchen Liu  


<div class="notes">
... slide notes ...
</div>


# Questions?


<div class="notes">
... slide notes ...
</div> 




## Shiny




<div class="notes">
... slide notes ...
</div>


## Riboviz Shiny output 




<div class="notes">
... slide notes ...
</div>



## Initial Attempts (Shiny)




<div class="notes">
... slide notes ...
</div>   

## The Solution (Shiny)




<div class="notes">
... slide notes ...
</div>
