---
title: "Lab Talk - Riboviz"
author: "Flic Anderson"
date: "10/05/2021"
output: 
  ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

<script>
    $(document).ready(function() {
      $('slide:not(.title-slide, .backdrop, .segue)').append('<footer label=\"bit.ly/Rmarkdown-and-output"></footer>');    
    })
</script>

<style>
  footer:after {
    content: attr(label);
    font-size: 12pt;
    position: absolute;
    bottom: 20px;
    left: 100px;
    line-height: 1.9;
  }
</style>

<style type="text/css">
slides > slide:not(.nobackground):after {
  content: '';
}
</style>


# TITLE: Workflow management systems & setting up Rmarkdown reports

## HOW TO (RE-)USE THIS MATERIAL

This is a `.html` presentation created in `R Markdown` with `ioslides`. 

(It's been written in a [.Rmd](http://rmarkdown.rstudio.com) file, and we generated .html slides by 'knitting' it in Rstudio.)

You can check out the code used to make these slides at the [Workshop repo](https://github.com/FlicAnderson/20201127-gitflow-workshop) on Github, and adapt it for your own presentations if you like - we've got a MIT Licence on the repo, which means: 

*"Basically, you can do whatever you want as long as you include the original copyright and license notice in any copy of the software/source."*

Source: [tl;drLegal](https://tldrlegal.com/license/mit-license)

# Outline: 

## Outline: 

* option
* option
* option

# Intro:

## RiboViz


<div class="notes">
... slide notes ...
</div>

## Ribosome Profiling Data

* Looking for 3-NT periodicity: ribosomes moving along transcript 1 codon at a time  

* Most reads map to coding regions (98.8% in Ingolia et al. 2009)  
* Reasonable read-lengths (e.g. know should look for appx ~28-30NT)  

* Looking for most reads to be in one frame

<div class="notes">
... slide notes ...
</div>

## Ribosome Profiling Method

<div class="centered">
```{r, out.width = "700"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/Ingolia2014_polysomes-to-footprints.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Inputs

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_inputs.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-aligned.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Processing

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_processing-ribogrid.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Analysis

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_analysis.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Outputs (NEW)

<div class="centered">
```{r, out.width = "800"}
#knitr::include_graphics("")
``` 
</div>

<div class="notes">
... slide notes ...
</div>


## Riboviz Workflow: Visualization (NEW)

<div class="centered">
```{r, out.width = "800"}
knitr::include_graphics("https://raw.githubusercontent.com/FlicAnderson/LabTalk-riboviz-20210510/main/images/riboviz-workflow_visualization.png")
``` 
</div>

<div class="notes">
... slide notes ...
</div>

## Workflow Management Systems



<div class="notes">
... slide notes ...
</div>



## Nextflow

* what is it
* why do we use it?
* what was the reasoning of switching to this from python/bash?


<div class="notes">
... slide notes ...
</div>


## Rmarkdown 




<div class="notes">
... slide notes ...
</div>



## Rmarkdown Outputs




<div class="notes">
... slide notes ...
</div>



## Shiny




<div class="notes">
... slide notes ...
</div>


## Riboviz HTML output 




<div class="notes">
... slide notes ...
</div>


## Riboviz Shiny output 




<div class="notes">
... slide notes ...
</div>


## Initial Attempts (HTML)




<div class="notes">
... slide notes ...
</div>


## Initial Attempts (Shiny)




<div class="notes">
... slide notes ...
</div>

## Next Attempts 




<div class="notes">
... slide notes ...
</div>


## The Solution (HTML)




<div class="notes">
... slide notes ...
</div>



## The Solution (Shiny)




<div class="notes">
... slide notes ...
</div>


## Summary / Lessons Learned




<div class="notes">
... slide notes ...
</div>


## Thanks/Acknowledgements

riboviz is a BBSRC-NSF funded collaborative project: 


* **The Wallace Lab** - *University of Edinburgh*:  
Edward Wallace, Emma MacKenzie, Sophie Winterbourne + past students (Ania Kurowska, Siyin Xue)  
* **The Shah Lab** -  *Rutgers University*:  
Premal Shah, John Favate, Alex Cope, Peter Tilton  
* **The Lareau Lab** - *University of California, Berkeley*:  
Liana Lareau, Amanda Mok, Vikram Shivakumar  
* **EPCC** - *University of Edinburgh*:  
Kostas Kavousannakis, Mike Jackson, Junchen Liu  


<div class="notes">
... slide notes ...
</div>


# Questions?


<div class="notes">
... slide notes ...
</div>